lib/AST/ASTPrinter.cpp

```diff
@@ -4043,15 +4069,489 @@ void PrintAST::visitMissingMemberDecl(MissingMemberDecl *decl) {
   Printer << " */";
 }
 
+void PrintAST::visitIntegerLiteralExpr(IntegerLiteralExpr *expr) {
+  Printer << expr->getDigitsText();
+}
+
+void PrintAST::visitFloatLiteralExpr(FloatLiteralExpr *expr) {
+  Printer << expr->getDigitsText();
+}
+
+void PrintAST::visitNilLiteralExpr(NilLiteralExpr *expr) {
+  Printer << "nil";
+}
+
+void PrintAST::visitStringLiteralExpr(StringLiteralExpr *expr) {
+  Printer << "\"" << expr->getValue() << "\"";
+}
+
+void PrintAST::visitBooleanLiteralExpr(BooleanLiteralExpr *expr) {
+  if (expr->getValue()) {
+    Printer << "true";
+  } else {
+    Printer << "false";
+  }
+}
+
+void PrintAST::visitRegexLiteralExpr(RegexLiteralExpr *expr) {
+  Printer << expr->getRegexText();
+}
+
+void PrintAST::visitErrorExpr(ErrorExpr *expr) {
+  Printer << "<error>";
+}
+
+void PrintAST::visitIfExpr(IfExpr *expr) {
+}
+
+void PrintAST::visitIsExpr(IsExpr *expr) {
+}
+
+void PrintAST::visitTapExpr(TapExpr *expr) {
+}
+
+void PrintAST::visitTryExpr(TryExpr *expr) {
+  Printer << "try ";
+  visit(expr->getSubExpr());
+}
+
+void PrintAST::visitCallExpr(CallExpr *expr) {
+  visit(expr->getFn());
+  Printer << "(";
+  auto argList = expr->getArgs();
+  bool isFirst = true;
+  // FIXME: handle trailing closures.
+  if (argList) {
+    for (auto i = argList->begin(), iEnd = argList->end(); i != iEnd; ++i) {
+      auto arg = *i;
+      if (dyn_cast<DefaultArgumentExpr>(arg.getExpr())) {
+        // Don't print default arguments.
+        continue;
+      }
+      if (!isFirst) {
+        Printer << ", ";
+      }
+      printArgument(arg);
+      isFirst = false;
+    }
+  }
+  Printer << ")";
+}
+
+void PrintAST::printArgument(const Argument &arg) {
+  auto label = arg.getLabel();
+  if (!label.empty()) {
+    Printer << label.str();
+    Printer << ": ";
+  }
+  if (arg.isInOut()) {
+    Printer << "&";
+  }
+  visit(arg.getExpr());
+}
+
+void PrintAST::visitLoadExpr(LoadExpr *expr) {
+  visit(expr->getSubExpr());
+}
+
+void PrintAST::visitTypeExpr(TypeExpr *expr) {
+  printType(expr->getType());
+}
+
+void PrintAST::visitArrayExpr(ArrayExpr *expr) {
+  Printer << "[";
+  bool isFirst = true;
+  auto elements = expr->getElements();
+  for (auto element : elements) {
+    if (!isFirst) {
+      Printer << ", ";
+    }
+    visit(element);
+    isFirst = false;
+  }
+  Printer << "]";
+}
+
+void PrintAST::visitDictionaryExpr(DictionaryExpr *expr) {
+  Printer << "[";
+  bool isFirst = true;
+  auto elements = expr->getElements();
+  for (auto element : elements) {
+    auto *tupleExpr = dyn_cast<TupleExpr>(element);
```

Comment:
I *believe* it's guaranteed for a `DictionaryExpr` to have `TupleExpr` elements, each with two elements, so I think you should be able to `cast<TupleExpr>` here, and access the key and value tuple elements directly with `->getElement` later. Really we ought to assert this in `DictionaryExpr`'s constructor (or better yet move it onto a better element representation).